File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/RegisterDeviceViewController.m
//
//  RegisterDeviceViewController.m
//  
//
//  Created by Federal Authority For Identity and Citizenship on 13/11/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "RegisterDeviceViewController.h"

@interface RegisterDeviceViewController ()
@property(nonatomic,strong) Utils * utils;
@end

@implementation RegisterDeviceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.deviceIDText.hidden=YES;
    self.deviceRegisIDText.hidden=YES;
    self.deviceRegisIDText.adjustsFontSizeToFitWidth=YES;

    self.title=@"Register Device";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(goBack)];
    
    self.registerDeviceButton.layer.cornerRadius=22.0;
    [self.registerDeviceButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.registerDeviceButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.registerDeviceButton.layer setShadowOpacity:0.5];
//    self.registerDeviceButton.layer.borderWidth=1.0;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    self.userIdText.text=@"";
    self.passwordText.text=@"";
    self.deviceRefText.text=@"";
    
    // Do any additional setup after loading the view.
}
- (IBAction)registerDeviceButtonAction:(id)sender {

    if ([self.userIdText.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_DEVICE_USERNAME onView:self ];
    } // if
    else if ([self.passwordText.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_DEVICE_PASSWORD onView:self];
    }//else if
    else if ([self.deviceRefText.text length]==0) {
        [AlertView showAlertTitle:ALERT withMessage:ENTER_DEVICE_REF_ID onView:self];
    }// else if
    else {
        [self.userIdText setUserInteractionEnabled:YES];
        [self.userIdText resignFirstResponder];
        [self.passwordText setUserInteractionEnabled:YES];
        [self.passwordText resignFirstResponder];
        [self.deviceRefText setUserInteractionEnabled:YES];
        [self.deviceRefText resignFirstResponder];

    [self.utils ShowProgressBar:LOADINGDATA andView:self.view];
    NSString *requestId=[Utils generateSecureKey];
    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(void){
    @try {
        Toolkit *toolkit = self.model.toolkitShared;
        DataProtectionKey *data_protection_key=self.model.dataprotectedShared;
        
        NSString *request_handle =  [toolkit prepareRequest:requestId];
        if (request_handle.length==0 || [request_handle isEqualToString:@""] || [request_handle isEqualToString:@"(null)"] ) {
            [AlertView showAlertTitle:ALERT withMessage:REQUEST_HANDLE_EMPTY onView:self];
        }//if
        else {
            NSString *userIdText=[Utils setEncrytion:request_handle data:self.userIdText.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
            if (userIdText.length==0 || [userIdText isEqualToString:@""] || [userIdText isEqualToString:@"(null)"] ) {
                [AlertView showAlertTitle:ALERT withMessage:ENCODE_USERID_EMPTY onView:self];
            }//if
            else {
                NSString *passwordText=[Utils setEncrytion:request_handle data:self.passwordText.text publickey:[data_protection_key getPublicKey] keylength:[data_protection_key getKeyLength]];
                if (passwordText.length==0 || [passwordText isEqualToString:@""] || [passwordText isEqualToString:@"(null)"] ) {
                    [AlertView showAlertTitle:ALERT withMessage:ENCODE_PASSWORD_EMPTY onView:self];
                }//if
                else {
                    RegisterDeviceResponse *registerDevice=[toolkit registerDevice:userIdText encodedPassword:passwordText deviceReferenceId:self.deviceRefText.text];
                    
                dispatch_async(dispatch_get_main_queue(), ^{
                    NSString *xmlString =[registerDevice getXmlString];
                    NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:xmlString];
                    if (![errorMsg isEqualToString:@""] || errorMsg.length>0) {
                        [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                    }//if
                    else {
                        NSString *deviceid = [registerDevice getDeviceRegistrationId];
                        self.deviceIDText.hidden=NO;
                        self.deviceRegisIDText.hidden=NO;
                        self.deviceRegisIDText.text=deviceid;
                        }//else
                     });// update UI main queue
                  }//else
               }//else
            }//else
        }//try
        @catch (NSException *exception) {
            NSString *vgresponse =exception.userInfo[@"ErrorResponse"];
            if (![vgresponse isEqualToString:@""] && vgresponse.length>0) {
                NSString *errorMsg =  [self.utils validateToolkitResponse:requestId xmlstring:vgresponse];
                if (![errorMsg isEqualToString:@""] && errorMsg.length>0) {
                    [AlertView showAlertTitle:ALERT withMessage:errorMsg onView:self];
                }//if
                else {
                    NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                    [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
                }//else
            }//if
            else {
                NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
                [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
            }
        }// catch
            @finally {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.utils DismissProgressBar];
                    self.userIdText.text=@"";
                    self.passwordText.text=@"";
                    self.deviceRefText.text=@"";
                });// update UI main queue
            }// finally
        });// background queue
    }//else
}
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    
     [self.userIdText setUserInteractionEnabled:YES];
     [self.userIdText resignFirstResponder];
     [self.passwordText setUserInteractionEnabled:YES];
     [self.passwordText resignFirstResponder];
     [self.deviceRefText setUserInteractionEnabled:YES];
     [self.deviceRefText resignFirstResponder];
    
    return YES;
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/RegisterDeviceViewController.swift
//
//  RegisterDeviceViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 22/04/22.
//

import UIKit

let utils = Utils()

class RegisterDeviceViewController: UIViewController,UITextFieldDelegate {
    
    @IBOutlet weak var userIdText: UITextField!
    @IBOutlet weak var passwordText: UITextField!
    @IBOutlet weak var deviceRefText: UITextField!
    @IBOutlet weak var deviceIDText: UILabel!
    @IBOutlet weak var deviceRegisIDText: UILabel!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        userIdText.delegate = self
        passwordText.delegate = self
        deviceRefText.delegate = self
        deviceIDText.isHidden = true
        
        
        deviceRegisIDText.isHidden = true
        deviceRegisIDText.adjustsFontSizeToFitWidth=true
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
    }
    
    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }    
    
    @IBAction func btn_RegisterDevice(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        
        if self.userIdText.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterDeviceUserName, onView: self)
        }//if
        else if self.passwordText.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterDevicePassword, onView: self)
        }//else if
        else if self.deviceRefText.text?.count==0 {
            AlertView.showAlertTitle(Constants.Alert, withMessage: Constants.EnterDeviceRefId, onView: self)
        }//else if
        else {
            userIdText.isUserInteractionEnabled = true
            userIdText.resignFirstResponder()
            passwordText.isUserInteractionEnabled = true
            passwordText.resignFirstResponder()
            deviceRefText.isUserInteractionEnabled = true
            deviceRefText.resignFirstResponder()
            
        utils.showProgressBar(Constants.LoadingData, andView: self.view)
        let requestId = Utils.generateSecureKey()
        DispatchQueue.global(qos: .background).async {
        do {
            let request_handle = try toolkit?.prepareRequest(requestId!)
            if (request_handle?.isEmpty  == true) || (request_handle == "") || (request_handle == "(null)") {
                AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Request_Handle_Empty, onView: self)
            }//if
            else {
                let en_userIdText:String? = Utils.setEncrytion(request_handle, data: self.userIdText.text!,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                if (en_userIdText?.isEmpty)! || (en_userIdText == "") || (en_userIdText == "(null)") {
                    AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_UserID_Empty, onView: self)
                }//if
                else {
                    let en_passwordText:String? = Utils.setEncrytion(request_handle, data:  self.passwordText.text!,publickey: dataProtectionKey.getPublicKey(), keylength: Int32(dataProtectionKey.getKeyLength()))
                    if (en_passwordText?.isEmpty)! || (en_passwordText == "") || (en_passwordText == "(null)") {
                         AlertView.showAlertTitle(Constants.Alert, withMessage:Constants.Encode_Password_Empty, onView: self)
                    }//if
                    else {
                registerDevice =   try  toolkit?.registerDevice(en_userIdText!, encodedPassword: en_passwordText!, deviceReferenceId: self.deviceRefText.text!)
                        DispatchQueue.main.async {
                            let deviceid: String = registerDevice.getDeviceRegistrationId()
                            self.deviceIDText.isHidden = false
                            self.deviceRegisIDText.isHidden = false
                            self.deviceRegisIDText.text = deviceid
                        }// update UI main queue
//                DispatchQueue.main.async {
//                let xmlString = registerDevice.getXmlString()
//                let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: xmlString)
//                if (errorMsg != "") && (errorMsg.count ) > 0 {
//                    AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
//                }//if
//                else {
//                    let deviceid: String = registerDevice.getDeviceRegistrationId()
//                    self.deviceIDText.isHidden = false
//                    self.deviceRegisIDText.isHidden = false
//                    self.deviceRegisIDText.text = deviceid
//                 }//else
//                }// update UI main queue
              }//else
             }//else
            }//else
            DispatchQueue.main.async {
                utils.dismissProgressBar()
            }// update UI main queue
        }//do
        catch let error as NSError {
            DispatchQueue.main.async {
               var vgresponse = String()
                if error.userInfo["ErrorResponse"] != nil {
                    vgresponse = error.userInfo["ErrorResponse"] as! String
                if (vgresponse != "") && ((vgresponse.count ) > 0) {
//                    let errorMsg: String = utils.validateToolkitResponse(requestId, xmlstring: vgresponse)
//                    if (errorMsg != "") && ((errorMsg.count ) > 0) {
//                        AlertView.showAlertTitle(Constants.Alert, withMessage:errorMsg, onView: self)
//                    }//if
//                    else {
                        let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                        AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
//                    }//else
                 }//if
                  else {
                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                    }//else
                }//if
                else {
                    let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                    AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                }//else
                utils.dismissProgressBar()
             }// update UI main queue
            }//catch
          }//background queue
        }//else
    }
    
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        userIdText.isUserInteractionEnabled = true
        userIdText.resignFirstResponder()
        passwordText.isUserInteractionEnabled = true
        passwordText.resignFirstResponder()
        deviceRefText.isUserInteractionEnabled = true
        deviceRefText.resignFirstResponder()
        
        return true
    }
    
    
}
